main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala

```diff
@@ -905,6 +906,194 @@ object CompletionProvider {
     }
   }
 
+  /**
+    * Gets completions after use keyword
+    */
+  private def getUseCompletions()(implicit context: Context, root: TypedAst.Root): Iterable[CompletionItem] = {
+    if (root == null) {
+      return Nil
+    }
+
+    val regex = raw"\s*use\s+(.*)".r
+
+    context.prefix match {
+      case regex(ns) => {
+        // Six cases
+        // 0: We have nothing i.e.
+        // 1: a path i.e. A/B
+        // 2: an item i.e. Foo (enum/class/def/type alias)
+        // 3: path and item i.e. A/B.Foo
+        // 4: item and tag/sig i.e Foo.Bar
+        // 5: path, item and tag/sig i.e A/B.Foo.Bar
+
+        val segments = ns.split('.');
+        segments.toList match { 
+          // case 0
+          case Nil => nsCompletionsAfterPrefix(Nil) ++ getItemUseCompletions(Nil)
+          // case 1/2
+          case x :: Nil => {
+            // We might be done typing the namespace or not. We need to try both cases
+            val prefix1 = x.split('/').toList;
+            val prefix2 = x.split('/').dropRight(1).toList;
+            // case 1
+            nsCompletionsAfterPrefix(prefix1) ++
+              nsCompletionsAfterPrefix(prefix2) ++ 
+              getItemUseCompletions(prefix1) ++
+              getItemUseCompletions(prefix2) ++
+              // case 2
+              getEnumTagCompletions(Nil, x) ++
+              getClassSigCompletions(Nil, x) 
+          }
+          // case 3/4
+          case x :: y :: Nil => {
+            val ns = x.split('/').toList;
+            // case 3
+            getItemUseCompletions(ns) ++
+              getEnumTagCompletions(ns, y) ++
+              getClassSigCompletions(ns, y) ++
+              // case 4
+              getEnumTagCompletions(Nil, x) ++
+              getClassSigCompletions(Nil, x)
+          }
+          // case 5
+          case x :: y :: _ :: Nil => {
+            val ns = x.split('/').toList;
+            getEnumTagCompletions(ns, y) ++
+              getClassSigCompletions(ns, y)
+          }
+          case _ => Nil
+        }
+      }
+      case _ => Nil
+    }
+  }
+
+  /**
+    * Gets completions for a sub namespace of a prefix namespace
+    * I.e if you have namespace A/B/C/D, then if prefix is A/B it will return a completion for A/B/C
+    */
+  private def nsCompletionsAfterPrefix(prefix: List[String])(implicit context: Context, root: TypedAst.Root): Iterable[CompletionItem] = {
+    val nss = root.defs.keySet.map(_.namespace) ++
```

Comment:
I guess this could be expensive, but at the same time, you don't have the right data structure to make it cheap.